core/src/main/java/zipkin2/reporter/BytesMessageSender.java

```diff
@@ -0,0 +1,116 @@
+ * also allows them to use their most natural message type. For example, kafka would more naturally
+ * send messages as byte arrays.
+ *
+ * @since 3.2
+ */
+public interface BytesMessageSender extends Closeable {
+
+  /**
+   * Base class for implementation, which implements {@link #messageSizeInBytes(List)} and
+   * {@link #messageSizeInBytes(List)} with a given {@linkplain Encoding}
+   */
+  abstract class Base implements BytesMessageSender {
+    protected final Encoding encoding;
+
+    protected Base(Encoding encoding) {
+      this.encoding = encoding;
+    }
+
+    /** {@inheritDoc} */
+    @Override public Encoding encoding() {
+      return encoding;
+    }
+
+    /** {@inheritDoc} */
+    @Override public int messageSizeInBytes(List<byte[]> encodedSpans) {
+      return encoding.listSizeInBytes(encodedSpans);
+    }
+
+    /** {@inheritDoc} */
+    @Override public int messageSizeInBytes(int encodedSizeInBytes) {
+      return encoding.listSizeInBytes(encodedSizeInBytes);
+    }
+  }
+
+  /** Returns the encoding this sender requires spans to have. */
+  Encoding encoding();
+
+  /**
+   * Maximum bytes sendable per message including overhead. This can be calculated using {@link
+   * #messageSizeInBytes(List)}
+   * <p>
+   * Defaults to 500KB as a conservative default. You may get better or reduced performance
+   * by changing this value based on, e.g., machine size or network bandwidth in your
+   * infrastructure. Finding a perfect value will require trying out different values in production,
+   * but the default should work well enough in most cases.
+   */
+  int messageMaxBytes();
+
+  /**
+   * Before invoking {@link BytesMessageSender#send(List)}, callers must consider message overhead,
+   * which might be more than encoding overhead. This is used to not exceed {@link
+   * BytesMessageSender#messageMaxBytes()}.
+   *
+   * <p>Note this is not always {@link Encoding#listSizeInBytes(List)}, as some senders have
+   * inefficient list encoding. For example, Scribe base64's then tags each span with a category.
+   */
+  int messageSizeInBytes(List<byte[]> encodedSpans);
+
+  /**
+   * Like {@link #messageSizeInBytes(List)}, except for a single-span. This is used to ensure a span
+   * is never accepted that can never be sent.
+   *
+   * <p>Note this is not always {@link Encoding#listSizeInBytes(int)}, as some senders have
+   * inefficient list encoding. For example, Stackdriver's proto message contains other fields.
+   *
+   * @param encodedSizeInBytes the {@link BytesEncoder#sizeInBytes(Object) encoded size} of a span
+   */
+  int messageSizeInBytes(int encodedSizeInBytes);
+
+  /**
+   * Sends a list of encoded spans to a transport such as HTTP or Kafka.
+   *
+   * @param encodedSpans list of encoded spans.
+   * @throws IllegalStateException if {@link #close() close} was called.
+   */
+  void send(List<byte[]> encodedSpans) throws IOException;
```

Comment:
(This is just an opinion) The semantics of sending an empty list as to way to very the connection could be surprising - logically I think noop is expected. May be having a dedicated check / ping method is not as bad after all

Reply:
I really regret having put this check in the first place as it was a relic from finagle failfast. It isn't used for anything, but it is totally the case that sending an empty list is permitted today. I would rather document an empty list is permitted since almost all impls of check had to send an empty list. This is better than forcing people to implement check which would only do the same, and then people think it is smarter than that.

Reply:
please have a look at the new RATIONALE.md for notes on this!

Reply:
Thanks a lot @codefromthecrypt for very comprehensive clarification! (knowing context always pays off)